README.md
# cgf-leapmotion-pd

## Environment

- MacOS 10.13.6 (윈도 10은 [이 문서](./docs/troubleshooting-for-win10.md)를 참고해주세요.)

- pd 0.49.1-i386

- node v12.13.0 or higher (comes with npm 6.12.0)


## Installation

0. 필요한 프로그램을 설치합니다.

    - Leap Motion SDK V2 for Mac: https://developer.leapmotion.com/setup/desktop

    - Node v12.13.0 이상 : https://nodejs.org/ko/download/releases/ 또는 https://nodejs.org/ko/download/

1. `Download ZIP` 또는 `git clone`을 통해 다운받습니다.

2. ZIP 파일의 압축을 해제합니다.

3. 터미널을 켠 후, 압축을 해제한 해당 폴더의 경로로 들어갑니다.

    - 에) 다운로드 폴더에 있을 경우 `cd ~/Downloads/cgf-leapmotion-pd`

4. `node/` 폴더 내에서 `npm install` 명령어를 입력합니다.


## Operation

1. Finder에서 소스 폴더의 경로로 들어갑니다.

2. `pd` 폴더에서 `node-osc-pd-02.pd` 또는 `node-osc-pd-03-processed.pd` 를 켭니다.

    - `node-osc-pd-02.pd` 와 `node-osc-pd-03-processed.pd` 는 손가락 서보의 파라미터를 조정하는 방법이 다릅니다.

3. 터미널을 켠 후, 소스 폴더의 경로로 들어갑니다.

4. `node/` 폴더 내에서 다음 명령어를 입력합니다.

    - `node-osc-pd-02.pd` 실행시
        
        `npm run start`

    - `node-osc-pd-03-processed.pd` 실행시

        `npm run processed-start`

## Tips

- [윈도 10 환경에서 Leap Motion 인식 안되는 문제 해결](./docs/troubleshooting-for-win10.md)


## Contact

김건호

gh.leokim@gmail.com


docs/history.md
# History

v0.1.0 : 초기 버전
- 링크 https://github.com/ghleokim/cgf-leapmotion-pd/tree/v0.1.0

v0.1.1 : 높낮이 정보 추가 (2020-01-12)
- 링크 https://github.com/ghleokim/cgf-leapmotion-pd/tree/v0.1.1

v0.1.2 : Development 모드 사용 가능. console log에 현재 프레임의 정보가 표시됩니다. (2020-08-03)
- 링크 https://github.com/ghleokim/cgf-leapmotion-pd/tree/v0.1.2


node/main.js
const math = require('./math.js');
const angle = math.angle;

const Leap = require('leapjs');
const {
  Client,
  Message
} = require('node-osc');

let timestamp = Date.now()

function onConnect() {
  console.log('Device Connected')
}

function frameCheck(frame) {
  if (frame.id % 10 === 0) {
    curTime = Date.now();
    timeDelta = curTime - timestamp;
    timestamp = curTime;

    console.log(`frame id : \x1b[36m${frame.id}\x1b[0m, \x1b[36m${timeDelta / 1000}\x1b[0m seconds every 10 frames`)
  }
}

const CustomFingers = function (data) {
  this.fingers = data
  this.bends = [0, 0, 0, 0, 0]
}

CustomFingers.prototype.getBends = function () {
  const options = {
    clip: {
      0: [0.25, 0.28, 0.3],
      1: [0.7, 0.6, 0.5],
      2: [0.7, 0.6, 0.5],
      3: [0.7, 0.6, 0.5],
      4: [0.7, 0.6, 0.5]
    },
    mix: {
      0: [0.3, 0, 0.7],
      1: [0.6, 0.3, 0.1],
      2: [0.6, 0.3, 0.1],
      3: [0.6, 0.3, 0.1],
      4: [0.6, 0.3, 0.1]
    }
  }

  this.bends = this.bends.map((value, index) => {
    const metArr = this.fingers[index].metacarpal.direction()
    const proxArr = this.fingers[index].proximal.direction()
    const medArr = this.fingers[index].medial.direction()
    const disArr = this.fingers[index].distal.direction()

    const bendItem = {
      metacarpal: metArr,
      proximal: proxArr,
      medial: medArr,
      distal: disArr,
      angle: [1 - angle(metArr, proxArr), 1 - angle(proxArr, medArr), 1 - angle(medArr, disArr)]
    }

    return bendItem
  })

  return this.bends
}

// console log out env mode : production or dev?
const __DEV__ = process.env.NODE_ENV === 'development'

console.log(`Hello, \x1b[33m clubgoldenflower \x1b[0m
Program is running in ${!!process.env.NODE_ENV ? process.env.NODE_ENV:'production' } mode.
`)

const controller = new Leap.Controller();

controller.on('connect', onConnect);

controller.loop(function (frame) {
  __DEV__ && frameCheck(frame)

  if (frame.hands[0]) {
    const a = new CustomFingers(frame.hands[0].fingers)
    const bends = a.getBends()
    const msg = new Message('/fingers/raw')
    msg.append('thumb')
    msg.append(bends[0].angle)
    msg.append('index')
    msg.append(bends[1].angle)
    msg.append('middle')
    msg.append(bends[2].angle)
    msg.append('ring')
    msg.append(bends[3].angle)
    msg.append('pinky')
    msg.append(bends[4].angle)

    const client = new Client('localhost', 3030)
    client.send(msg, (err) => {
      if (err) {
        console.log(err)
      }
      client.close()
    })

    // process wrist and roll
    const roll = (-1 * frame.hands[0].roll() + Math.PI) / (2 * Math.PI)
    const palm = frame.hands[0].palmNormal
    const arm = frame.hands[0].arm.direction()
    const wrist = (angle(palm, arm) + 1) / 2

    const msg2 = new Message('/hand')
    msg2.append('s6_wrist')
    msg2.append(wrist)
    msg2.append('s7_roll')
    msg2.append(roll)

    const palmPos = frame.hands[0].palmPosition

    // added
    msg2.append('s8_')
    msg2.append(palmPos[1])
    msg2.append('s9_')
    msg2.append(palmPos[2])
    msg2.append('s10_')
    msg2.append(palmPos[0])

    const client2 = new Client('localhost', 3030)
    client2.send(msg2, (err) => {
      if (err) {
        console.log(err)
      }
      client2.close()
    })
  }

})

node/test2.js
function multiply(a, b) {
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
  }

function angle(a, b) {
    this.normalizedA = 0
    this.normalizedB = 0

    a.forEach(function (el) {
        this.normalizedA += el * el
    })
    b.forEach(function (el) {
        this.normalizedB += el * el
    })

    normalizedA = Math.sqrt(normalizedA)
    normalizedB = Math.sqrt(normalizedB)

    return (multiply(a, b) / (normalizedA * normalizedB))
}

const Leap = require('leapjs');
const controller = new Leap.Controller();

controller.loop(function (frame) {

            if (frame.hands[0]) {
                const palm = frame.hands[0].palmNormal
                const arm = frame.hands[0].arm.direction()
                
                console.log(palm)
                console.log(arm)
                console.log(angle(palm, arm))

                console.log('\n')
            }
})

node/jsproj-nodeosc.js
// make print functions
const readline = require('readline')
const blank = '\n'.repeat(process.stdout.rows)

function logger(msg) {
  process.stdout.write(blank)
  readline.cursorTo(process.stdout, 0, 0)
  process.stdout.write(msg)
}

function multiply(a, b) {
  let res = 0

  res += a[0] * b[0];
  res += a[1] * b[1];
  res += a[2] * b[2];

  return res
}

function angle(a, b) {
  this.normalizedA = 0
  this.normalizedB = 0

  a.forEach(function (el) {
    this.normalizedA += el * el
  })
  b.forEach(function (el) {
    this.normalizedB += el * el
  })

  normalizedA = Math.sqrt(normalizedA)
  normalizedB = Math.sqrt(normalizedB)

  // console.log(normalizedA, normalizedB)

  return Math.acos(multiply(a, b) / (normalizedA * normalizedB))
}


// start
const {
  Client,
  Message
} = require('node-osc');

let client = new Client('localhost', 3000);

client.send('/test', 10, () => {
  client.close()
})

const Leap = require('leapjs');
const controller = new Leap.Controller();

//finger

function fingerElement (finger) {
  const metArr = finger.metacarpal.direction()
  const proxArr = finger.proximal.direction()
  const medArr = finger.medial.direction()

  return [ metArr, proxArr, medArr ]
}

controller.loop(function (frame) {
  // console.log(state)
  // console.log(frame.hands[0])
  if (frame.hands[0]) {
    const fingers = []

    frame.hands[0].fingers.forEach(function (el) {
      const [ metArr, proxArr, medArr ] = fingerElement(el)
      // console.log(metArr, proxArr)
      fingers.push([angle(metArr, proxArr), angle(proxArr, medArr)])
    })

    const msg = `${fingers[0]}\n${fingers[1]}\n${fingers[2]}\n${fingers[3]}\n${fingers[4]}`

    // console.log(fingers)
    logger(msg)

    const testFinger = frame.hands[0].fingers[0]

    const metArr = testFinger.metacarpal.direction()
    const proxArr = testFinger.proximal.direction()
    const medArr = testFinger.medial.direction()

    // const msg = testFinger.stabilizedTipPosition.forEach((e) => {
    //     parseInt(e)
    //   }) +
    //   `\nmatrix: ${testFinger.metacarpal.matrix()}` +
    //   `\nmetacarpal : ${metArr}` +
    //   `\nproximal : ${proxArr}` +
    //   `\nmedial : ${medArr}` +
    //   `\ncos met-prox : ${angle(metArr, proxArr)}` +
    //   `\ncos prox-med : ${angle(proxArr, medArr)}`
    // // + `\nangle : `

    // logger(msg)


    // basic frame
    const msg_frame = new Message('/frame')
    // msg_frame.append('valid')
    // msg_frame.append(frame.hands[0]['valid'])
    msg_frame.append('id')
    msg_frame.append(frame.hands[0].id)

    msg_frame.append('palmNormal')
    msg_frame.append(frame.hands[0].palmNormal)

    let client = new Client('localhost', 3000);
    client.send(msg_frame, (err) => {
      if (err) {
        console.log(err)
      }
      client.close()
    })

    msg_finger0 = new Message('/finger')
    msg_finger0.append('metacarpal')
    msg_finger0.append(metArr)
    msg_finger0.append('proximal')
    msg_finger0.append(proxArr)
    msg_finger0.append('medial')
    msg_finger0.append(medArr)

    client.send(msg_finger0, (err) => {
      if (err) {
        console.log(err)
      }
      client.close()
    })
  }
})

node/main-processed.js
const math = require('./math.js');
const angle = math.angle;

const Leap = require('leapjs');
const {
  Client,
  Message
} = require('node-osc');

let timestamp = Date.now()

function onConnect() {
  console.log('Device Connected')
}

function frameCheck(frame) {
  if (frame.id % 10 === 0) {
    curTime = Date.now();
    timeDelta = curTime - timestamp;
    timestamp = curTime;

    console.log(`frame id : \x1b[36m${frame.id}\x1b[0m, \x1b[36m${timeDelta / 1000}\x1b[0m seconds every 10 frames`)
  }
}

function processAngle(clip, mix, angle) {

  angle = angle.map((value, index) => {
    return Math.min(clip[index], value) / clip[index]
  })
  return angle[0] * mix[0] + angle[1] * mix[1] + angle[2] * mix[2]

}

const fs = require('fs')
const content = fs.readFileSync("options.json")
const jsonContent = JSON.parse(content)

const CustomFingers = function (data) {
  this.fingers = data
  this.bends = [0, 0, 0, 0, 0]
}

CustomFingers.prototype.getBends = function () {
  // clip and mix for each joints in fingers
  // get options from options.json
  const options = jsonContent

  this.bends = this.bends.map((value, index) => {
    const metArr = this.fingers[index].metacarpal.direction()
    const proxArr = this.fingers[index].proximal.direction()
    const medArr = this.fingers[index].medial.direction()
    const disArr = this.fingers[index].distal.direction()

    const bendItem = {
      metacarpal: metArr,
      proximal: proxArr,
      medial: medArr,
      distal: disArr,
      angle: [1 - angle(metArr, proxArr), 1 - angle(proxArr, medArr), 1 - angle(medArr, disArr)],
      processed: processAngle(
        options.clip[index], options.mix[index],
        [1 - angle(metArr, proxArr), 1 - angle(proxArr, medArr), 1 - angle(medArr, disArr)]
      )
    }

    return bendItem
  })

  return this.bends
}

// console log out env mode : production or dev?
const __DEV__ = process.env.NODE_ENV === 'development'

console.log(`Hello, \x1b[33m clubgoldenflower \x1b[0m
Program is running in ${!!process.env.NODE_ENV ? process.env.NODE_ENV:'production' } mode.
`)

const controller = new Leap.Controller();

controller.on('connect', onConnect);

controller.loop(function (frame) {
  __DEV__ && frameCheck(frame)

  if (frame.hands[0] && frame.id % 8 == 1) {
    const a = new CustomFingers(frame.hands[0].fingers)
    const bends = a.getBends()

    const msg = new Message('/fingers/processed')
    msg.append('thumb')
    msg.append(bends[0].processed)
    msg.append('index')
    msg.append(bends[1].processed)
    msg.append('middle')
    msg.append(bends[2].processed)
    msg.append('ring')
    msg.append(bends[3].processed)
    msg.append('pinky')
    msg.append(bends[4].processed)

    const client = new Client('localhost', 3000)
    client.send(msg, (err) => {
      if (err) {
        console.log(err)
      }
      client.close()
    })

    // process wrist and roll
    const roll = (-1 * frame.hands[0].roll() + Math.PI) / (2 * Math.PI)
    const palm = frame.hands[0].palmNormal
    const arm = frame.hands[0].arm.direction()
    const wrist = (angle(palm, arm) + 1) / 2

    const msg2 = new Message('/hand')
    msg2.append('s6_wrist')
    msg2.append(wrist)
    msg2.append('s7_roll')
    msg2.append(roll)

    const palmPos = frame.hands[0].palmPosition

    // added
    msg2.append('s8_')
    msg2.append(palmPos[1])
    msg2.append('s9_')
    msg2.append(palmPos[2])
    msg2.append('s10_')
    msg2.append(palmPos[0])

    const client2 = new Client('localhost', 3000)
    client2.send(msg2, (err) => {
      if (err) {
        console.log(err)
      }
      client2.close()
    })
  }

})